cpp/src/arrow/engine/substrait/options.cc

```diff
@@ -166,6 +171,57 @@ class DefaultExtensionProvider : public BaseExtensionProvider {
                                      named_tap_rel.name(), std::move(renamed_schema)));
     return RelationInfo{{std::move(decl), std::move(renamed_schema)}, std::nullopt};
   }
+
+  Result<RelationInfo> MakeSegmentedAggregateRel(
+      const ConversionOptions& conv_opts, const std::vector<DeclarationInfo>& inputs,
+      const substrait_ext::SegmentedAggregateRel& seg_agg_rel,
+      const ExtensionSet& ext_set) {
+    if (inputs.size() != 1) {
+      return Status::Invalid(
+          "substrait_ext::SegmentedAggregateRel requires a single input but got: ",
+          inputs.size());
+    }
+
+    auto input_schema = inputs[0].output_schema;
+
+    ConversionOptions conversion_options;
+
+    // store segment key fields to be used when output schema is created
+    std::vector<int> segment_key_field_ids;
+    std::vector<FieldRef> segment_keys;
+    if (seg_agg_rel.segment_groupings_size() > 0) {
+      ARROW_RETURN_NOT_OK(internal::ParseAggregateGrouping(
+          seg_agg_rel.segment_groupings(0), ext_set, conversion_options, input_schema,
+          &segment_key_field_ids, &segment_keys));
+    }
+
+    const auto& aggregate = seg_agg_rel.aggregate();
+    ARROW_ASSIGN_OR_RAISE(
+        auto decl_info,
+        internal::ParseAggregateDeclaration(
```

Comment:
I am kind of between here, if we copy the AggregationRel and turn that into a new SegmentedAggregationRel, then I don't know if there is a reasonable way to reuse the deser code for AggregationRel  (or do we need to copy the code)?

If we use this approach, most of the deser code can be shared but the rel message looks a bit weird. So I am not sure which way of better.

@westonpace WDYT?

Reply:
IIRC, we discussed a couple of alternatives and leaned towards the copying one; another alternative was reusing `AggregateRel` in the message and leveraging the existing parsing code for, and while working on the PR I found that this was simpler to implement.

Reply:
In Substrait itself we have been discouraging this kind of approach when creating physical relations because:

 * It's too expressive - We don't consume all kinds of AggregateRel (e.g. expressions in a grouping have to be direct references) and, since this is a physical relation, we should only expose what we can consume.
 * Unnecessary coupling - It's not possible to change AggregateRel without potentially changing all the extensions and it's not clear they would always need to change.
 * Directly including the Rel itself leads to some awkwardness like the fact that you now have multiple "inputs".

However, yes, most of the parsing code would then have to be duplicated.  So for something internal I don't know that it is unworkable.  So I don't have a strong opinion but I would lean slightly towards something like...

```
message SegmentedAggregateRel {
  repeated Expression.FieldReference grouping_keys = 0;
  repeated Expression.FieldReference segment_keys = 1;
  repeated substrait.AggregateRel.Measure measures = 2;
}
```

Reply:
@icexelloss, do we have time now to try the approach Weston suggests?

Reply:
Yes I think we should do what Weston suggests

Reply:
Fixed.